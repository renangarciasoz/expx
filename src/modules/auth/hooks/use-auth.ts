import { api } from "api/api-client";
import {
  recoverUserInformation,
  signInRequest,
  SignInRequestData,
} from "api/auth.api";
import { AxiosError } from "axios";
import { EXPIRATION_IN_HOURS, TOKEN_NAME } from "constants/auth.constants";
import { DASHBOARD, HOME } from "constants/urls.constants";
import { useRouter } from "next/router";
import { destroyCookie, parseCookies, setCookie } from "nookies";
import { useEffect } from "react";
import { useStore } from "store/auth.store";

export function useAuth() {
  const router = useRouter();
  const { user, error, setUser, setError, removeUser, ...store } = useStore();
  const isAuthenticated = !!user;

  useEffect(() => {
    const { [TOKEN_NAME]: token } = parseCookies();

    if (token) {
      recoverUserInformation().then((user) => {
        setUser(user);
      });
    }
  }, [setUser]);

  async function signIn({ email, password }: SignInRequestData) {
    setError("");

    try {
      const { token, ...user } = await signInRequest({
        email,
        password,
      });

      setCookie(undefined, TOKEN_NAME, token, {
        maxAge: 60 * 60 * EXPIRATION_IN_HOURS, // seconds * minutes * constant number in hours
      });

      api.defaults.headers.common["Authorization"] = `Bearer ${token}`;

      setUser(user);

      router.push(DASHBOARD);
    } catch (e) {
      removeUser();
      delete api.defaults.headers.common["Authorization"];
      setError((e as AxiosError).message);
    }
  }

  async function signOut() {
    destroyCookie(undefined, TOKEN_NAME);
    delete api.defaults.headers.common["Authorization"];
    removeUser();
    router.push(HOME);
  }

  return {
    ...store,
    isAuthenticated,
    user,
    error,
    signIn,
    setError,
    signOut,
  };
}
