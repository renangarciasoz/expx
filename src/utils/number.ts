import { ONE_BILLION } from "constants/numbers.constants";

export const formatValue = (value: number, minimumFractionDigits = 2) =>
  Intl.NumberFormat("en", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits,
  }).format(value);

export const formatBillionValue = (
  value: string | number,
  minimumFractionDigits = 3
) => {
  return `${formatValue(Number(value) / ONE_BILLION, minimumFractionDigits)}b`;
};
