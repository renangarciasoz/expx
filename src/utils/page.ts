export const blockPageScroll = () => {
  const body = document?.querySelector("body");
  body!.style.height = "100vh";
  body!.style.overflow = "hidden";
};

export const unblockPageScroll = () => {
  const body = document?.querySelector("body");
  body!.style.height = "auto";
  body!.style.overflow = "auto";
};
