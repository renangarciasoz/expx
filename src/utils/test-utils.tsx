import { render, RenderOptions } from "@testing-library/react"; //jest-ignore
import { i18n } from "configs/i18next-tests.config";
import { ThemeProvider } from "contexts/theme.ctx";
import React, { ReactElement } from "react";
import { I18nextProvider } from "react-i18next";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </ThemeProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
