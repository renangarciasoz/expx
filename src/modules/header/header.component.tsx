import { ArrowCircleDown, ArrowCircleUp } from "@mui/icons-material";
import { AppBar, Box, IconButton, Toolbar, Typography } from "@mui/material";
import { COMPANY_NAME } from "constants/company.constants";
import { POSITION_TOP } from "constants/header.constants";
import { AuthBtns } from "modules/auth/auth-btns.component";
import { AuthForm } from "modules/auth/auth-form.component";
import { useHeader } from "modules/header/hooks/use-header";

export const Header = () => {
  const { authFormIsOpened, headerIsAtTopPosition, changeHeaderPosition } =
    useHeader(); // ?

  return (
    <AppBar
      data-testid="header"
      position="fixed"
      sx={{
        top: headerIsAtTopPosition ? 0 : "unset",
        bottom: headerIsAtTopPosition ? "unset" : 20,
        flexDirection: headerIsAtTopPosition ? "column" : "column-reverse",
        height: authFormIsOpened ? "100vh" : "auto",
        display: "flex",
        backgroundColor: "grey.1000",
        p: 1,
      }}
      elevation={0}
    >
      <Toolbar sx={{ boxShadow: "none" }}>
        <Typography
          data-testid="logo"
          sx={{ flex: 1 }}
          variant="h4"
          component="h1"
        >
          {COMPANY_NAME}
        </Typography>
        {!authFormIsOpened && (
          <Box flex={1} display="flex" justifyContent="center">
            <IconButton
              data-testid="change-header-position-btn"
              onClick={() =>
                changeHeaderPosition(
                  headerIsAtTopPosition ? "bottom" : POSITION_TOP
                )
              }
            >
              {headerIsAtTopPosition ? <ArrowCircleDown /> : <ArrowCircleUp />}
            </IconButton>
          </Box>
        )}
        <AuthBtns />
      </Toolbar>
      {authFormIsOpened && <AuthForm />}
    </AppBar>
  );
};
