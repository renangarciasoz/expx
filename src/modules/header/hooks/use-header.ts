import { POSITION_TOP } from "constants/header.constants";
import { useStore as useAuthStore } from "store/auth.store";
import { useStore } from "store/header.store";
import { blockPageScroll, unblockPageScroll } from "utils/page";

export const useHeader = () => {
  const { position, setFormAuthIsOpened, ...store } = useStore();
  const { setError } = useAuthStore();

  const headerIsAtTopPosition = position === POSITION_TOP;

  function openAuthForm() {
    blockPageScroll();
    setFormAuthIsOpened(true);
  }

  function closeAuthForm() {
    unblockPageScroll();
    setFormAuthIsOpened(false);
    setError(null);
  }

  return {
    openAuthForm,
    closeAuthForm,
    headerIsAtTopPosition,
    ...store,
  };
};
