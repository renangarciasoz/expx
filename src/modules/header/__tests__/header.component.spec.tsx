/**
 * @jest-environment jsdom
 */

import { fireEvent, render, screen } from "utils/test-utils";
import { Header } from "../header.component";
import * as useHeader from "../hooks/use-header";

jest.mock("modules/auth/auth-form.component", () => ({
  AuthForm: () => <div />,
}));
jest.mock("modules/auth/auth-btns.component", () => ({
  AuthBtns: () => <div data-testid="auth-form" />,
}));

describe("Header", () => {
  beforeEach(() => {
    render(<Header />);
  });

  it("renders correctly", () => {
    expect(document.body).toMatchSnapshot();
    expect(screen.getByTestId("logo")).toBeInTheDocument();
  });

  it("changes toolbar position to bottom and to top again", () => {
    const changeToolbarPositionBtn = screen.getByTestId(
      "change-header-position-btn"
    );

    fireEvent.click(changeToolbarPositionBtn);
    const headerBottomStyles = window.getComputedStyle(
      screen.getByTestId("header")
    );
    expect(headerBottomStyles.bottom).toBe("20px");
    expect(headerBottomStyles.top).toBe("");
    expect(screen.getByTestId("ArrowCircleUpIcon")).toBeInTheDocument();

    fireEvent.click(changeToolbarPositionBtn);
    const headerTopStyles = window.getComputedStyle(
      screen.getByTestId("header")
    );
    expect(headerTopStyles.bottom).toBe("");
    expect(headerTopStyles.top).toBe("0px");
    expect(screen.getByTestId("ArrowCircleDownIcon")).toBeInTheDocument();
  });

  describe("when auth form is opened", () => {
    beforeAll(() => {
      jest.spyOn(useHeader, "useHeader").mockImplementation(() => ({
        ...jest.requireActual("modules/header/hooks/use-header"),
        authFormIsOpened: true,
      }));
    });

    it("header is fullscreen", () => {
      const headerStyles = window.getComputedStyle(
        screen.getByTestId("header")
      );
      expect(headerStyles.height).toBe("100vh");
    });

    it("renders auth form", () => {
      expect(screen.getByTestId("auth-form")).toBeInTheDocument();
    });
  });
});
