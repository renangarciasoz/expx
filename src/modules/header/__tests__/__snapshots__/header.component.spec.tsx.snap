// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-10d26i6-MuiPaper-root-MuiAppBar-root"
      data-testid="header"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1prgphj-MuiToolbar-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4 css-1ht2c9-MuiTypography-root"
          data-testid="logo"
        >
          ExPx
        </h1>
        <div
          class="MuiBox-root css-t508yj"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="change-header-position-btn"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowCircleDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 10V8h-2v4H8l4 4 4-4h-3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          data-testid="auth-form"
        />
      </div>
    </header>
  </div>
</body>
`;
